### ISSUE TITLE
Add ampersand support for selectors

### ISSUE THREAD

**Issue by jjenzz** [modulz/stitches#60]
Our primitives would export something like:

```jsx
export const resetStyles = {
  root: {
    display: 'flex',

    '&[data-state-disabled]': {
      pointerEvents: 'none',
    },
  },
};
```

and then the consumer would do:

```jsx
import { ComponentPrimitive, resetStyles } from '@interop/primitives';

const Component = styled(ComponentPrimitive, {
  ...resetStyles.root,
  border: '1px solid green';
});
```

The ampersand means this spreading by the consumer would work in styled-components, emotion and also stitches.

@christianalfoni was thinking we could have selector logic as:

- If it starts with `&`, remove it
- If it now starts with `:` or `[`, treat as state selector
- If not, insert a space to treat as nested element selector

```jsx
{
  '&:hover': {},
  'div': {},
  '&[..]': {},
}
```

**Comment by jjenzz** [comment]
Also, I raised a comment before about potentially interpolating a class from a component `toString` method https://github.com/modulz/stitches/issues/52#issuecomment-656708402. 

Specifically the ability to do something like:

```jsx
[`${Input}:checked + &`]: {
  color: 'green',
  border: '1px solid red',
}
```

If our `Input.toString()` returned a randomly generated class specific to the component (`.i1234`) could we maybe make this output the following?

```jsx
.i1234:checked + .color1 { color: 'green'; }
.i1234:checked + .border-size1 { border-size: '1px' }
.i1234:checked + .border-style1 { border-style: 'solid' }
.i1234:checked + .border-color1 { border-color: 'red' }
```

In this case, the `&` wouldn't be removed and instead it would be replaced with the atomic class.

**Comment by peduarte** [comment]
@hadihallak maybe a good first issue? @christianalfoni can review

**Comment by peduarte** [comment]
Initial implementation here https://github.com/modulz/stitches/pull/88

We only support `&` at the _**beginning of the selector**_. This is mainly for familiarity with other devs and remove the need to have a `space` to target descendants.

**Comment by christianalfoni** [comment]
Okay, so I did an implementation of this on `poc` for reference.

Instead of looking at the `prop` to figure out if it is a selector or not, I rather look at `props[prop]`, if it is an object it means there are nested styles, which means the `prop` is a selector

This makes things a lot more straight forward it looks. Basically creating the whole selector now becomes:

```ts
    // Allow using "&" to position the classname, also multiple times
    const selector =
      this.pseudo && this.pseudo.includes("&")
        ? this.pseudo.replace(/&/gi, `.${className}`)
        : this.pseudo
        ? `.${className}${this.pseudo}`
        : `.${className}`;
```

**Comment by hadihallak** [comment]
Fixed in #112
